tests/models/gemma/test_modeling_gemma.py

```diff
@@ -831,6 +832,67 @@ def test_compile_static_cache(self):
         static_compiled_text = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
         self.assertEqual(EXPECTED_TEXT_COMPLETION, static_compiled_text)
 
+    @slow
+    @require_read_token
+    def test_export_static_cache(self):
+        if version.parse(torch.__version__) < version.parse("2.3.0"):
+            self.skipTest(reason="This test requires torch >= 2.3 to run.")
+
+        from transformers.integrations.executorch import (
+            TorchExportableModuleWithStaticCache,
+            convert_and_export_with_cache,
+        )
+
+        tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b", pad_token="</s>", padding_side="right")
+        EXPECTED_TEXT_COMPLETION = [
+            "Hello I am doing a project on the 1990s and I need to know what the most popular music was in the 1990s. I have looked on the internet and I have found a few things but I need more. I have found that the most popular music was rap",
```

Comment:
@ArthurZucker I see a test failure may be related to this. I copied this over from another model with additional texts added, feel free to truncate it to a shorter message and push a new commit

Reply:
Ah yeah won't have time to update them tonight! But yeah we can merge otherwise!

Reply:
@ArthurZucker Let me know if there is anything I can do to merge this PR

Reply:
I will check this afternoon and merge

Reply:
gemma: all passing

gemma2: `test_export_static_cache`

I am getting the following (the output is the same as the prompt), and I am not sure that is expected.

```
(Pdb) ep_generated_text
['Hello I am doing']
(Pdb) ep_generated_ids
tensor([[   2, 4521,  590, 1144, 3900,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0]])

```

Reply:
For gemma, it could generate 

```
(Pdb) ep_generated_text
['Hello I am doing a project on the 1990s and I need to know what the most popular music was in the 1990s. I have looked on the internet and I have found']

```

Reply:
@ydshieh thank you for helping on merging this PR. Are you running the test on CPU or GPU? Actually the export test is expected to run on CPU only, however, the entire `Gemma2IntegrationTest` is decorated `@require_torch_gpu`. To verify the correctness locally, if you comment out `@require_torch_gpu`, then run `RUN_SLOW=1 pytest tests/models/gemma2/test_modeling_gemma2.py -k test_export_static_cache`, it shows
```
tests/models/gemma2/test_modeling_gemma2.py::Gemma2IntegrationTest::test_export_static_cache huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:01<00:00,  1.98it/s]
PASSED
======================================================== 1 passed, 388 deselected, 4 warnings in 25.10s ========================================================
```

Reply:
@ydshieh Yes, happy to explain more. The exported program is an intermediate artifact that should not be directly used for product or anything. Instead users typically should take this intermediate artifact and further lower it to ExecuTorch which will covers the hardware delegates, op fusion, memory planning, etc. So the starting point (i.e. export) is always on CPU. This is different from the torch compiled artifacts. So the reason I'm adding a `generate` util in this PR is to validate the correctness of the exported program and ensure it can be further consumed in the downstream ExecuTorch stack. This is clearly documented in the util API itself. So for this test the perf doesn't matter, only correctness maters, so we don't expect to run this test on GPU. And it may produce weird output even if it may work in some cases.

Reply:
I thought it may be okay to run the test on a machine configured with GPU because the `device=cpu` is already hard-coded in the test itself, maybe I'm wrong. I'm wondering if there is a tag to mark this test cpu-only?

Reply:
Thank you @ydshieh ! Once this PR is merged I can start scaling the model coverage to other models utilizing the same util function added in this PR.

Reply:
Could you share your torch version ..?

Reply:
I finally get it pass (on our runner) by changing to

> attn_implementation = None

No idea why `sdpa` always get the original prompt as output.

Reply:
I will push a commit and merge tomorrow morning.

Reply:
> Could you share your torch version ..?

Mine is a dev version `2.6.0.dev20241007`. What is the torch version where the test fails?

>  attn_implementation = None

It doesn't look correct. If we set it to `None`, what is the default attention being used? If I recall correctly, SPDA is the only attention impl that supports StaticCache.

Reply:
Yeah, torch version seem to be the issue. Let me dig into it and will update here shortly

Reply:
Okay, I can confirm that exporting `gemma2` model will require `torch==2.5.0` in order to work correctly.
I also verified that running the test on `torch==2.4.1` or `torch==2.0.0` will also get the original prompt as output.

Here are the detailed package info:
```
pip list | grep torch
executorch                         0.5.0a0+f8cec53
executorchcoreml                   0.0.1
torch                              2.5.0
torchaudio                         2.5.0
torchvision                        0.20.0
```

`RUN_SLOW=1 pytest tests/models/gemma2/test_modeling_gemma2.py -k test_export_static_cache -v`
```
=========================================================================================== test session starts ===========================================================================================
platform darwin -- Python 3.10.13, pytest-7.2.0, pluggy-1.0.0 -- /Users/guangyang/miniconda3/envs/executorch/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/guangyang/transformers/.hypothesis/examples')
rootdir: /Users/guangyang/transformers, configfile: pyproject.toml
plugins: cov-4.1.0, anyio-4.4.0, xdist-3.3.1, hypothesis-6.84.2
collected 389 items / 388 deselected / 1 selected

tests/models/gemma2/test_modeling_gemma2.py::Gemma2IntegrationTest::test_export_static_cache PASSED                                                                                                 [100%]
```

Reply:
I updated the PR to bump up the required `torch` version to `2.5.0`. 

`torch==2.5.0` is not available publicly until Oct 17, 2024. We can either merge this PR as-is, the `test_export_static_cache` for `gemma2` will be skipped in the CI until `torch 2.5.0` is available in `transformers`. Or we can defer `gemma2` enablement until `torch 2.5.0` is available in `transformers` by splitting the `gemma2` related code out of this PR and merge the rest. Personally I would prefer former, but let me know how you want to proceed. cc: @ArthurZucker @ydshieh

Reply:
Great! We are still using python 3.8 and torch 2.4.1. I plan to switch to python 3.9 at the end of October. Before that, python 2.5 won't be available in the system. We will think about that.

We can merge this PR as it is. Thank you for checking (it works indeed).

> If we set it to None, what is the default attention being used?

Just the native implementation of attention (in the modeling files) using torch operators.

> If I recall correctly, SPDA is the only attention impl that supports StaticCache.

The native implementation of attention also works with `StaticCache` I believe.